### DISCUSSION TITLE
Models, and the effect of these on the Frame Rate - Continued from Discussion #614

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#617] [General]
Extended from #614
The 614 discussion centered around the Time and Step in SK and how these could be determined, calculated, used, etc. That discussion led to practical use of code in the Step where time/performance is critical. This criticality exists due to the number of, and complexity of models used in industrial applications.
The last question spawned the suggestion to move this to another discussion thread.

The question was anchoring text to "objects" so that they can respond to SK inputs and provide text information to users based on their use in the application. My issue was the problem of placing the text in relation to the object it is describing, e.g. the length of a pipe being "gazed" upon by a user. Or in another context where the objects rotate to face the user and must move the text to accompany the objects where the text appears between the object and the user, while the user walk around the object.

Here is the previous/last part of that discussion.

>Here's an example of a Model that always faces the user, with some text anchored to it! I've done the anchoring via matrices here, but you could also do this with the Hierarchy class too.
>In this particular sample, the GLTF file could contain a node called "TextAnchor", and if that was present, we would just use its transform for the text. If it's not present, it just uses an offset of (0,1,0). If the GLTF has scaling in it though, this could end out a bit weird, but it's not to hard to derive a scale-less transform from the existing node transform.
```csharp
// Probably cache these
Model     model      = Model.FromFile("asset.glb");
ModelNode textAnchor = model?.FindNode("TextAnchor");

// On Step
Vec3   modelPos       = Vec3.Zero;
Matrix modelTransform = Matrix.TR(modelPos, Quat.LookAt(modelPos, Input.Head.position));
Matrix textTransform  = textAnchor?.ModelTransform ?? Matrix.T(0,1,0);

model.Draw(modelTransform);
Text.Add("Info here", textTransform*modelTransform);
```
This suggestion uses a model from a file. I am creating the "model" procedurally in code by creating a hierarchical model starting with a root node with and ID and name, otherwise empty. Then I would add child nodes to the model with their own ID and name and other properties.  My assumption is that the model nodes could be created where the internal to the model spatial relationship is preserved. Example: Make a cube with a rotation defined such that it faces the user. This means that some side of the cube is the "facing side" and the rest are not. I want to construct a cube and have text positioned between the user and the cube. I want the text to orbit around the cube's center such that it is always between the user and the cube. From the example code above it would seem that the object and the text can be anchored together "in the model".
Trying to figure out this myself is a learning experience so in the SK documentation found directly under Model, I come upon this paragraph: 

> Because models include an offset transform for each mesh element, this does have the overhead of an extra matrix multiplication in order to execute a render command. So if you need speed, and only have a single mesh with a precalculated transform matrix, it can be faster to render a Mesh instead of a Model!

Does this mean that all model are subject to the extra matrix multiplication or does this only apply to procedurally generated models?

**Comment by maluoi** [comment]
> This suggestion uses a model from a file. I am creating the "model" procedurally in code by creating a hierarchical model starting with a root node with and ID and name, otherwise empty.

A procedurally assembled `Model` is functionally the same as one loaded from a file. The file loading code uses the same functions as you do to assemble a `Model` via API.

> My assumption is that the model nodes could be created where the internal to the model spatial relationship is preserved.

A `Model` is like a small scene. When a `ModelNode` is a parent of another `ModelNode`, there is a spatial relationship, child nodes are always relative to their parent node. A `ModelNode` has a [`LocalTransform`](https://stereokit.net/Pages/StereoKit/ModelNode/LocalTransform.html)(tranform relative to its own parent) and a [`ModelTransform`](https://stereokit.net/Pages/StereoKit/ModelNode/ModelTransform.html)(transform relative to the `Model`'s origin). Updating either one of these transforms will _also_ update the `ModelTransform` of all their child nodes to preserve their local relationship.

> Example: Make a cube with a rotation defined such that it faces the user. This means that some side of the cube is the "facing side" and the rest are not. I want to construct a cube and have text positioned between the user and the cube. I want the text to orbit around the cube's center such that it is always between the user and the cube. From the example code above it would seem that the object and the text can be anchored together "in the model".

If you want to rotate the whole `Model`, then the above code will do exactly this for you :) Text isn't a `Mesh` in the traditional sense, so you can't assign it to a `Model` in the same way. So for that, we just need to make sure the text shows up in the right place by using the ModelNode's info!

If you want a `ModelNode` to face you instead of the whole model, you'll need to play a little with spaces.

```csharp
// Probably cache these
Model     model = Model.FromFile("asset.glb");
ModelNode node  = model.FindNode("FacingNode");
Matrix    modelTransform = Matrix.T(1,2,3);

// On Step

// Take the head position, and move it into the Model's coordinate space.
Vec3 modelSpaceHead = modelTransform.Inverse.Transform(Input.Head.position);

// Make the node look at the head by doing all our work in Model space
node.ModelTransform = Matrix.TRS(
    node,ModelTransform.Translation,
    Quat.LookDir(modelSpaceHead,
    node.ModelTransform.Scale));

// Whatever offset we want for the text from the ModelNode
Matrix textTransform = Matrix.T(0,1,0);

model.Draw(modelTransform);
Text.Add("Info here", textTransform*node.ModelTransform);
```

> Does this mean that all model are subject to the extra matrix multiplication or does this only apply to procedurally generated models?

`Model`s internal data structure is essentially just a collection of `Mesh`/`Material` pairs with a transform relative to the `Model` origin! When you call `model.Draw(at)`, this is essentially what happens under the hood (pseudocode):

```csharp
foreach (NodeData n in model.VisualNodes) {
    n.Mesh.Draw(n.Material, n.ModelTransform * at);
}
```

So you do have an extra matrix multiply in there, since the `Model` can't know in advance where it will get drawn. If you _do_ have a completely static set of `Model`s, then you can cache some of that information as `Mesh`/`Material`/`Matrix`(world transform) and skip the extra Matrix call. You can see [something quite similar in the light baking sample here](https://github.com/maluoi/SKLightBake/blob/main/StaticScene.cs).

That being said, this is **_not_** a big cost in small to medium scenes and I would _not_ suggest that you try to make this optimization until you have profiled your application and decided that it is definitely a bottleneck. Bottlenecks in XR are typically on the GPU, not CPU. Caching things does open up the codebase to caching bugs.

**Comment by laultman** [comment]
"In the same coordinate space" or similar phrases. I think my largest problem with the spatial relationship of object is the phrase used often in the documentation is not easily grasped nor is it explicitly explained with examples. I get the part about Window and Handle as magic containers where the relationship is obvious. I don't seem to follow the external connection between the head or hands being somehow "in the same coordinate space" or how you do this. I grasp the fact that they all need to use the same base coordinates or even locals if just among themselves. It would be helpful to have documentation devoted to the theory and mechanics of this because it is critical part of the SK drawing system.

My view of relationships is not what is happening. For example the placing of text between the user and a model, the model is tracking the head because in this code:
```csharp
bool grabActive = UI.HandleBegin(Id, ref _poseHandle, _poseBounds, false, _grabUiMoveOption);
// Make the target face the user 
Quat boxRot = Quat.LookAt(_poseHandle.position, Input.Head.position);
_boxMesh.Draw(Default.MaterialUI, Matrix.TRS(_position, boxRot, _size), _currentTargetColor);
model.Model.Draw(Matrix.TRS(_position, boxRot, _size), _currentTargetColor);
```
So, it's not the head looking at the model, it's the model looking at the head? If that is true, then the ray cast cannot give you a distance between the origins of the two. I was trying in my mind to make a ray from the head to the model origin, then find a point along the ray to place the text in the Step. You seem to be saying that if they are all in the same space then they are drawn as a unit, which makes sense with this new to me perspective.

**Reply by maluoi** [reply]
Coordinate space stuff can be tricky! They still mess me up quite regularly.

When you have an xyz position, say <1,1,0>, you know this is one unit over on the x axis, and one unit up the y axis. ...But relative to what, and in what units? Relative to the Sun? The earth? The the application's origin? Your head? One up and one over must be relative to something else! Every position and orientation is in some "coordinate space".

In StereoKit, when we say "World Space", this means "Relative to the application's origin, in meters" which in Real World Space is generally where the user's head was when the app started up. When we say "Model Space", this means "Relative to the `Model`'s origin", which in World Space is wherever you decided to `Draw()` the `Model`.

When you have position `A`<1,1,0> in World Space and position `B`<1,1,0> in Model Space, The distance between them is most likely **_not_** zero. You can't actually do math on these positions as they are! It's meaningless since they're in totally different coordinate spaces! To find the distance, they must both be in the same coordinate space, relative to the same origin point. This could be moving `A` into Model Space, or moving `B` into World Space, depending on the context. It's also good to note that if the Model Space transform involves a scaling value other than 1, the distance would be different in World Space vs. Model Space, so which space you use might depend on if you wanted World Space distance or Model Space distance!

If you're using `Hierarchy`, like with `Window` or `Handle`, you can pretty easily move things to and from World Space to Hierarchy Local Space using `Hierarchy.ToLocal` and `Hierarchy.ToWorld`.

For Model Space, you'll need to work directly with a Matrix, but it's also pretty easy.

```csharp
Matrix modelTransform = Matrix.TRS(...);
mode.Draw(modelTransform);

Vec3 modelPoint = new Vec3(...);

// This moves the point from Model Space to World Space
Vec3 worldPoint = modelTransform.Transform(modelPoint);

// This moves the point from World Space right back to Model Space
modelPoint = modelTransform.Inverse.Transform(worldPoint);
```

**Comment by laultman** [comment]
**Update: See next comment after this one.**

I am getting closer to the solution. I have the model defined and it is tracking the head position but with a weird side effect. I can't seem to understand what is really happening. But quickly the model is "looking" over my head and behind me. I start off with a defined pose for the Handle that contains the model and some other handy items.
```	
	Vec3 _size = new Vec3(3, 3, 3) * U.cm;
	Pose _poseHandle = new Pose(0, 0, -.8f, Quat.Identity);
	Pose _poseHandleText = new Pose(0, 0, -.7f, Quat.Identity);
	Pose _modelPose = new Pose(0, 0, -.8f, Quat.Identity);
	ModelNode cubeAnchor;
	ModelNode textAnchor;
	StereoKit.Model model;
	Material mat;
	Material matxt;
```
Then in my Initialize() I define a simple Model to test my code.
```
	mat = Default.Material.Copy();  // make a copy to set properties as needed
	mat.SetColor(AppConstants.shaderColorParamName, _skColor);
	matxt = Default.Material.Copy();  // make a copy to set properties as needed
	matxt.SetColor(AppConstants.shaderColorParamName, StereoKit.Color.White);
	model = new StereoKit.Model("EyeGazeTrack", _boxMesh, mat);
	cubeAnchor = model?.FindNode("EyeGazeTrack");
	textAnchor = model?.FindNode("cubeText");
```
Next in my Step() I draw the Model in the Handle.
```
	bool grabActive = UI.HandleBegin(Id, ref _poseHandle, _poseBounds, false, _grabUiMoveOption);
	// Make the target face the user 
	Quat boxRot = Quat.LookAt(_modelPose.position, Input.Head.position);
	Matrix modelTransform = Matrix.TRS(_modelPose.position, boxRot, _size);
	Matrix textTransform = textAnchor?.ModelTransform ?? Matrix.T(0, 1, 0);
	model.Draw(modelTransform);
```
The value of the _poseHandle.position remains unchanged while the Input.Head.position is changing. In the very first call to Step the Head.position is as expected 0,0,0 and the pose is unchanged. On the very next scan the Head.position has moved a smidgen .08,.08.-.01 
Everything is drawing as expected. On the first frame everything seems to work. Further down in the code in the Step, I am using the Input.Eyes.Ray.Intersect method to test if I am looking at the model. I have a bounding box for that purpose that is centered on the model.
Inside the Handle's Pose I have this statement:
```
	if (Input.Eyes.Ray.Intersect(_boundsEyesArea, out Vec3 at))		// gaze focus on target?
	{
		sphereMesh.Draw(Default.Material, Matrix.TS(at, _size*2), _currentTargetColor);
	} 
```
When running, the model is properly placed out in front of me -.8 **BUT** it seems as though I have to look down a pipe to intersect with the model's bounding box. In fact, if I use my hand ray cast, it intersects some invisible object about 18 inches in front of my starting head position. If I maintain a hand ray on the invisible object as I move about, it is the invisible object that is the intersecting point, not the model. I cannot yet determine why this is happening. I suspect the Input.Eyes.Ray.Intersect is somehow not in the coordinate space and I don't know how to fix this.

**Comment by laultman** [comment]
I have run more test. The "invisible" object is the Handle's bounds object. I added a cube at the Handle position so I could see what is happening. The problem seems to be my needing to place multiple models in space that are relative to the Handle origin yet are at their own pose positions'.

**Comment by laultman** [comment]
**Update: I have a workable solution. Lot of experimentation using the DemoText sample and a brain refresh on Trig and basic Matrix math.**

I guess I am now back to looking for suggestions. Note: Movable to me is like a Handle object. Say, I want to build a collection of movable model objects (ie Handle) and I want to treat the whole bunch as a single movable object. A warehouse contains lots of pallets and pallets contain lots of models and so on...
I want to be able to move the warehouse (there can be many warehouses at a large industrial site.) I thus want all the stuff in a warehouse to move with it. I need to move things in the warehouse (ie pallets) and I need to unpack a pallet as individual movable items.
My thought was to create a hierarchy of Handle based things. Seemed to work ok. Then I needed to use Eye gaze to select individual objects. That seemed to work too until I added the second object. I am having trouble with the gaze being related to the Handle and not the model. Related to seems to mean in SK terminology, "in the same coordinate space".
I am at a loss of where to start. Hierarchy seems to bind things together but doesn't have a handle. Wide open to suggestions on how to start and approach the problem?

**Comment by laultman** [comment]
TL;DR Skip to the end to see the code to put text on an object and rotate them in tandem based on the Head input...

Object of the exercise, develop an industrial UI where users have both hands involved in work and safety is a major concern where blocked views of equipment could result in serious injury to a user.
I previously created a "menu" AKA action triggers (AT) that incorporate Eye/Gaze detection, selection, and dwell locked triggers. I needed a way for users to place the ATs in suitable locations in their workspace but it needed to default to some order of workflow for the job the user is tasked to do.
I decided on a semi-circular array of ATs above head height that becomes visible when the user looks up from his workstation. For example, a worker may need to summon a forklift when a container is full, a model object of a forklift as an AT could trigger our shopfloor management system to notify a forklift driver.
The problems were many, but one was most difficult, users had difficulty maintaining a gaze-lock on a target for a sufficiently long period to trigger the action. Noise, alerts, sharp sounds (clangs, bangs, crashes) cause users to glance away from the target. For safety and to cancel an action the "glance-away" immediately stops the action and locks out the action for a short period.
Solution to this glance-away turned out to be human engineering, give the user something to focus on that is moving (animated) and the human vision system kicks in to lock in on the moving item even with other things in the work area. Simple as it seems a countdown timer for the AT was sufficient.
The AT needed an associated text countdown that is inline with the Eye/Gaze intersection area so that the user's eyes remained focused within the detection zone of the AT. For this, the target (forklift) needed to rotate itself and the associated text for the user to easily maintain the focus. The ATs need to be aligned with the user's head position, unobstructed.
The main solution was to create a hierarchy object that contains and rotates both the AT and the text to face the user. As most things turn out to be, with SK it's not that difficult once you understand the "in the same space" phrase.

This code uses a Handle as the container for the model (mesh in this case) and the text to face the user. As the user moves the text faces the user. Constraints, the Handle pose is placed at about arm's length so that hand inputs are available and that it is easy to see with normal vision.
I'm now enhancing the code to have multiple ATs in a hierarchical "cluster".
Here is the code:
```
	Quat boxRot = Quat.LookAt(_poseHandle.position, Input.Head.position);	// capture the current rotation
	_poseHandle.orientation = boxRot;                                       // apply the rotation to the pose (rotating the entire hierarchy's space)

	// apply the rotation to the pose (rotating the entire hierarchy's space)
	bool grabActive = UI.HandleBegin("0", ref _poseHandle, _poseBounds, false, UIMove.PosOnly);
	_boxMesh.Draw(Default.Material, Matrix.TS(_position, _size), _skColor); // mesh is a cube sized 3,3,3 * U.cm

	// Adjust the position of the text to be between user and the model
	var meshZ = _boxMesh.Bounds.dimensions.z * _size.z;	// scale the bounds, we only need the z here
	float zL = meshZ / 2 + .01f;						// calculate the depth Z of the mesh and add a little to move it out in front of the object
	Vec3 txtPos = _position;							// _position in this case is Vec3.Zero, we are not moving it with respect to the Handle pose
	txtPos.z = -zL + txtPos.z;							// shift the text position closer to the user
				
	// draw the text, note alignment is centered which centers it with the model
	StereoKit.Text.Add("ME!", Matrix.T(txtPos), TextAlign.Center, alignX | alignY); // the text is centered on origin and alined left and top

	UI.HandleEnd();
```
There are several ways to implement this, code it in the step, create an object that can be registered with the step, create a derived class and instantiate it at runtime as a step module.
